Sources/src/FromSTMTestBeamData_source.cc

```diff
@@ -0,0 +1,335 @@
+//
+// Source module to read binary files from the STM test beam
+// at gELBE collected in April 2022
+//
+#include <iostream>
+#include <fstream>
+#include <boost/utility.hpp>
+#include <cassert>
+#include <set>
+#include <string>
+#include <vector>
+
+#include "fhiclcpp/types/Name.h"
+#include "art/Framework/IO/Sources/Source.h"
+#include "art/Framework/Core/InputSourceMacros.h"
+#include "art/Framework/IO/Sources/SourceHelper.h"
+#include "art/Framework/Principal/RunPrincipal.h"
+#include "art/Framework/Principal/SubRunPrincipal.h"
+#include "art/Framework/Principal/EventPrincipal.h"
+#include "art/Framework/IO/Sources/put_product_in_principal.h"
+#include "canvas/Persistency/Provenance/Timestamp.h"
+#include "canvas/Persistency/Provenance/RunID.h"
+#include "canvas/Persistency/Provenance/SubRunID.h"
+#include "canvas/Persistency/Provenance/EventID.h"
+#include "canvas/Persistency/Provenance/BranchType.h"
+#include "canvas/Persistency/Provenance/ProductID.h"
+#include "canvas/Persistency/Provenance/canonicalProductName.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Services/Registry/ServiceHandle.h"
+
+#include "Offline/Sources/inc/STMTestBeamHeaders.hh"
+#include "Offline/DataProducts/inc/STMTestBeamEventInfo.hh"
+#include "Offline/RecoDataProducts/inc/STMWaveformDigi.hh"
+
+
+using namespace std;
+
+namespace mu2e {
+  struct Config
+  {
+    using Name = fhicl::Name;
+    using Comment = fhicl::Comment;
+    fhicl::Sequence<std::string> inputFiles{Name("fileNames"),Comment("Input binary file")};
+    fhicl::Atom<unsigned int> maxEvents{Name("maxEvents"), Comment("Max number of events")};
+    fhicl::Atom<int> verbosityLevel{Name("verbosityLevel"), Comment("Verbosity level")};
+
+    // These are used by art and are required.
+    fhicl::Atom<std::string> module_label{Name("module_label"), Comment("Art module label"), ""};
+    fhicl::Atom<std::string> module_type{Name("module_type"), Comment("Art module type"), ""};
+  };
+  typedef fhicl::WrappedTable<Config> Parameters;
+
+
+  //================================================================
+  class STMTestBeamDataDetail : private boost::noncopyable {
+    std::string myModuleLabel_;
+    art::SourceHelper const& pm_;
+    unsigned runNumber_; // from file name
+    art::SubRunID lastSubRunID_;
+    std::set<art::SubRunID> seenSRIDs_;
+
+    unsigned currentFileNumber_;
+    std::string currentFileName_;
+    std::ifstream* currentFile_ = nullptr;
```

Comment:
You are doing everything right with this bare pointer.  I strongly recommend changing the type to
``std::unique_ptr<std:ifstream>``` .   That way a future maintainer does not need to think about calls to close or calling delete.  You can just reset the unique_ptr and both are done.  Or if you copy into the unique_ptr, the object being replaced has it's d'tor called.  That automatically calls close.   Basically this future proofs the code.

I imagine you followed a model in which it was done with bare pointers.  We should fix them all as targets of opportunity.

Reply:
Done here [2d0e936](https://github.com/Mu2e/Offline/pull/856/commits/2d0e936dcc79cfd39f811ed2a179cbede517c07f)